File: zzsong-iam-server/src/main/kotlin/com/zzsong/iam/server/infrastructure/encoder/password/IamPasswordEncoder.java
package com.zzsong.iam.server.infrastructure.encoder.password;

import cn.idealframework.crypto.HmacSHA1;
import cn.idealframework.crypto.MD5;
import cn.idealframework.crypto.SHA256;
import cn.idealframework.lang.StringUtils;
import org.springframework.stereotype.Component;

import javax.annotation.Nonnull;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

/**
 * @author 宋志宗 on 2021/11/20
 */
@Component
public class IamPasswordEncoder implements PasswordEncoder {
  private static final String PASSWORD_SALT = "u9*BUkjG*EFm4*M-x4LA7QnxdfK6PFdR";

  @Override
  public String encode(@Nonnull CharSequence rawPassword) {
    return encryptPwd(rawPassword);
  }

  @Override
  public boolean matches(@Nonnull CharSequence rawPassword, @Nonnull String encodedPassword) {
    return encryptPwd(rawPassword.toString()).equals(encodedPassword);
  }

  @Nonnull
  private String encryptPwd(@Nonnull CharSequence rawPassword) {
    String rawPasswordStr = rawPassword.toString();
    Base64.Encoder base64Encoder = Base64.getEncoder();
    String en1 = SHA256.encode(rawPasswordStr + PASSWORD_SALT);
    String en2 = base64Encoder.encodeToString(HmacSHA1.encode(PASSWORD_SALT, rawPasswordStr));
    String en3 = MD5.encode(en1 + PASSWORD_SALT + en2 + rawPasswordStr);
    String en3Base64 = base64Encoder.encodeToString(en3.getBytes(StandardCharsets.US_ASCII));
    String en1Base64 = base64Encoder.encodeToString(en1.getBytes(StandardCharsets.US_ASCII));
    String s = "$2a$" + en3Base64 + en1Base64 + en2;
    return StringUtils.replace(s, "=", "");
  }
}

File: zzsong-iam-user/src/main/kotlin/com/zzsong/iam/user/infrastructure/password/encoder/IamPasswordEncoder.kt
package com.zzsong.iam.user.infrastructure.password.encoder

import cn.idealframework.crypto.HmacSHA1
import cn.idealframework.crypto.MD5
import cn.idealframework.crypto.SHA256
import cn.idealframework.lang.StringUtils
import java.nio.charset.StandardCharsets
import java.util.*

/**
 * @author 宋志宗 on 2021/11/20
 */
object IamPasswordEncoder : PasswordEncoder {
  private const val SALT = "7aRw.fXAF*ohWgdJRH!aZyE7LEhWZ6Nz"

  override fun encode(rawPassword: CharSequence): String {
    return encryptPwd(rawPassword)
  }

  override fun matches(rawPassword: CharSequence, encodedPassword: String): Boolean {
    return encryptPwd(rawPassword.toString()) == encodedPassword
  }

  private fun encryptPwd(rawPassword: CharSequence): String {
    val rawPasswordStr = rawPassword.toString()
    val base64Encoder = Base64.getEncoder()
    val en1 = SHA256.encode(rawPasswordStr + SALT)
    val en2 = base64Encoder.encodeToString(HmacSHA1.encode(SALT, rawPasswordStr))
    val en3 = MD5.encode(en1 + SALT + en2 + rawPasswordStr)
    val en3Base64 = base64Encoder.encodeToString(en3.toByteArray(StandardCharsets.US_ASCII))
    val en1Base64 = base64Encoder.encodeToString(en1.toByteArray(StandardCharsets.US_ASCII))
    val s = "$2a$$en3Base64$en1Base64$en2"
    return StringUtils.replace(s, "=", "")
  }
}
